firmware/quadruna/BMS/src/app/app_accumulator.c

```diff
@@ -71,181 +58,301 @@ typedef enum
+    OWCFaults owcFaults = { .owc_status = { 0U }, .owc_fault_gnd = { 0U }, .owc_global_fault = 0U };
 
-    target_voltage = app_canRx_Debug_CellBalancingOverrideTarget_get()
-                         ? app_canRx_Debug_CellBalancingOverrideTargetValue_get()
-                         : target_voltage;
-    for (uint8_t segment = 0U; segment < ACCUMULATOR_NUM_SEGMENTS; segment++)
+    owcFaults.owc_global_fault = io_ltc6813CellVoltages_getGlobalOpenWireFault();
+
+    if (owcFaults.owc_global_fault)
     {
-        for (uint8_t cell = 0U; cell < ACCUMULATOR_NUM_SERIES_CELLS_PER_SEGMENT; cell++)
+        for (uint8_t segment = 0; segment < NUM_SEGMENTS; segment++)
         {
-            const bool needs_discharging = (io_ltc6813CellVoltages_getCellVoltage(segment, cell) > target_voltage);
-            io_ltc6813Shared_markCellForDischarge(segment, cell, needs_discharging);
+            if (io_ltc6813CellVoltages_getOpenWireFault(segment, 0))
+            {
+                owcFaults.owc_fault_gnd[segment] = true;
+                owcFaults.owc_status[segment]    = (uint16_t)1;
+            }
+            else
+            {
+                for (uint8_t cell = 1; cell < CELLS_PER_SEGMENT; cell++)
+                {
+                    if (io_ltc6813CellVoltages_getOpenWireFault(segment, cell))
+                    {
+                        owcFaults.owc_status[segment] |= ((uint16_t)(1 << cell));
+                    }
+                }
+            }
         }
     }
+
+    data.owc_faults = owcFaults;
 }
 
-void app_accumulator_balanceCells(void)
+static void calculateCellsToBalance(bool cells_to_balance[NUM_SEGMENTS][CELLS_PER_SEGMENT])
 {
-    if (!data.balance_enabled)
+    const float target_voltage = app_canRx_Debug_CellBalancingOverrideTarget_get()
+                                     ? app_canRx_Debug_CellBalancingOverrideTargetValue_get()
+                                     : voltage_stats.min_voltage_cell.voltage + CELL_VOLTAGE_BALANCE_WINDOW_V;
+    for (uint8_t segment = 0U; segment < NUM_SEGMENTS; segment++)
     {
-        io_ltc6813Shared_disableBalance();
-        return;
+        for (uint8_t cell = 0U; cell < CELLS_PER_SEGMENT; cell++)
+        {
+            cells_to_balance[segment][cell] = cell_voltages[segment][cell] > target_voltage;
+        }
     }
+}
 
+void app_accumulator_balanceCells(void)
+{
     // Exit early if ADC conversion fails
-    if (!io_ltc6813Shared_pollAdcConversions())
+    if (!io_ltc6813_pollAdcConversions())
     {
         return;
     }
 
     // Write to configuration register to configure cell discharging
-    io_ltc6813Shared_writeConfigurationRegisters(true);
+    bool cells_to_balance[NUM_SEGMENTS][CELLS_PER_SEGMENT];
+    calculateCellsToBalance(cells_to_balance);
+    io_ltc6813_writeConfigurationRegisters((LTCConfig){ .balance_config = &cells_to_balance });
 
+// Discharging cells continuously generates too much heat.
```

Comment:
Not sure what future plans are, but if I recall correctly this might not be true on quadruna, you can probably be more aggressive but that’s a nitpick